design/tls-backend-verification.md

```diff
@@ -0,0 +1,166 @@
+This was rejected for two reasons
+
+1. Contour already watches Secret objects, so we get this for free without having to watch a new set of objects.
+2. Using ConfigMaps creates a precident for storing other information in ConfigMaps. As ConfigMaps are just homeless annotations, their potential for misuse is endless.
+
+## Security Considerations
+
+I have some concerns about the efficacy of this mechanism to deliver _verification_ not just _encryption_.
+To explain my concern, the way that certificate validation occurs on the public internet is there are two third parties, the CA and the DNS host, both are known to the client and the server but neither are under the direct control of either.
+The client looks up the DNS name that it intends to connect too, receives a set of IP addresses, connects to those and compares that the certificate presented is both signed by a CA that it trusts, and carries a subjectAltName that matches the DNS name it originally connected too.
+There are a set of compensating controls at work
+
+a. The DNS record may be hijacked, but without the matching certificate this is of little use.
+
+b. The certificate may be stolen, but the attacker must also redirect the DNS entries
+
+c. A certificate may be signed by an alternative CA, which is why pinning certificate fingerprints is a thing.
+
+The probability of a & b happening at the same time is non zero, but represent at least some effort to compensate for the trust issues of each component, this is why c is a thing.
+ 
+Let's compare this to how verification would work inside a Kubernetes cluster:
+Envoy, acting as the client, is configured to talk to a set of IP addresses obtained from k8s Endpoint objects--there are no hostnames in play inside a cluster--representing pods for that service.
+Envoy is configured by Contour to perform a TLS handshake when connecting to a pod.
+This design will add the following constraints to Envoy's TLS handshake.
+
+a. That the certificate presented is signed by the CA who's public key we present to Envoy via the IngressRoute `spec.routes[].backends[]verification.caSecret` key.
+
+b. That the certificate presented contains a subjectAltName that is present in the IngressRoute `spec.routes[].backends[].verification.subjectName` key.
+
+The problem is that both of these checks exist _inside the security boundary of the application which we are attempting to validate_.
+Because the certificate is almost certainly going to be signed by a company CA, not one of the public CAs, checking the CA presented by the server matches the material supplied to the client, Envoy, is little more than a shared secret.
+It does not assert that the certificate _is_ signed by a trusted CA, only that the certificate _is_ signed by a key to which the client has a matching public component.
+This is further weakened by the observation that the service which we are connecting too, the ingressroute, and the caSecret are in the _same namespace_.
```

Comment:
I don't know if this is a design choice of Contour, but there's no technical reason ingressroutes couldn't be changed to refer to services in other namespaces.